File: apps/backend/app/schemas/skin_test.py
from pydantic import BaseModel, Field

from app.schemas.profile import SkinProfileData


class SkinTestOptionResponse(BaseModel):
    id: int
    text: str


class SkinTestQuestionResponse(BaseModel):
    id: int
    text: str
    options: list[SkinTestOptionResponse]
    skip_conditions: dict | list | None = None


class SkinTestQuestionsResponse(BaseModel):
    version: str
    questions: list[SkinTestQuestionResponse]


class SkinTestAnswerInput(BaseModel):
    question_id: int
    option_id: int


class SkinTestSubmitRequest(BaseModel):
    version: str | None = None
    anonymous_id: str | None = None
    answers: list[SkinTestAnswerInput] = Field(min_length=1)


class SkinTestResultData(BaseModel):
    result_id: int
    skin_type: str
    sensitivity: str
    recommended_effects: list[str] = Field(default_factory=list)
    avoid_hint: list[str] = Field(default_factory=list)
    concern_tags: list[str] = Field(default_factory=list)
    type_code: str
    title: str
    subtitle: str | None = None
    image_storage_key: str | None = None


class SkinTestResultResponse(BaseModel):
    result: SkinTestResultData


class SkinTestApplyRequest(BaseModel):
    result_id: int


class SkinTestApplyResponse(BaseModel):
    success: bool = True
    skin_profile: SkinProfileData | None = None

File: apps/frontend/src/types/skinTest.ts
export type SkinTestTypeCode =
  | "OSNT"
  | "OSNW"
  | "OSPT"
  | "OSPW"
  | "ORNT"
  | "ORNW"
  | "ORPT"
  | "ORPW"
  | "DSNT"
  | "DSNW"
  | "DSPT"
  | "DSPW"
  | "DRNT"
  | "DRNW"
  | "DRPT"
  | "DRPW";

export type SkinTestOption = {
  id: number | string;
  text: string;
};

export type SkinTestQuestion = {
  id: number | string;
  text: string;
  options: SkinTestOption[];
  skip_conditions?: unknown;
};

export type SkinTestQuestionsResponse = {
  version: string;
  questions: SkinTestQuestion[];
};

export type SkinTestAnswer = {
  question_id: number | string;
  option_id: number | string;
};

export type SkinTestSubmitRequest = {
  version: string;
  answers: SkinTestAnswer[];
};

export type SkinTestResult = {
  result_id: number;
  skin_type: string;
  sensitivity: string;
  recommended_effects: string[];
  avoid_hint?: string[];
  concern_tags?: string[];
  type_code?: SkinTestTypeCode | string;
  title?: string;
  subtitle?: string;
  image_storage_key?: string;
};

export type SkinTestSubmitResponse = SkinTestResult;

export type SkinTestResultResponse = {
  result: SkinTestResult;
};

export type ApplySkinTestResultRequest = {
  result_id: number;
};

export type ApplySkinTestResultResponse = {
  success?: boolean;
  skin_profile?: {
    skin_type?: string | null;
    sensitivity?: string | null;
    latest_skin_test_result_id?: number | null;
  };
};
